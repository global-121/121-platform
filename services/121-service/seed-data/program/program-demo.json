{
  "published": false,
  "validation": true,
  "phase": "design",
  "location": "Demoland",
  "ngo": "Demo NGO",
  "titlePortal": {
    "en": "Demo Cash Aid Project"
  },
  "titlePaApp": {
    "en": "Demo Cash Aid Project"
  },
  "description": {
    "en": ""
  },
  "startDate": "2022-09-29T12:00:00Z",
  "endDate": "2030-11-30T12:00:00Z",
  "currency": "EUR",
  "distributionFrequency": "week",
  "distributionDuration": 10,
  "fixedTransferValue": 35,
  "financialServiceProviders": [
    {
      "fsp": "Intersolve-voucher-whatsapp",
      "configuration": [
        {
          "name": "username",
          "value": "INTERSOLVE_USERNAME"
        },
        {
          "name": "password",
          "value": "INTERSOLVE_PASSWORD"
        }
      ]
    },
    {
      "fsp": "Africas-talking"
    }
  ],
  "targetNrRegistrations": 870,
  "tryWhatsAppFirst": true,
  "meetingDocuments": {
    "en": ""
  },
  "phoneNumberPlaceholder": "+31 600 00 00 00",
  "programCustomAttributes": [],
  "programQuestions": [
    {
      "name": "name",
      "label": {
        "en": "1. What is your name?"
      },
      "answerType": "text",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Name"
      }
    },
    {
      "name": "middlename",
      "label": {
        "en": "2. What is your middle name?"
      },
      "answerType": "text",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Middle name"
      }
    },
    {
      "name": "firstname",
      "label": {
        "en": "3. What is your first name?"
      },
      "answerType": "text",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "First name"
      }
    },
    {
      "name": "gender",
      "label": {
        "en": "4. What is your gender?"
      },
      "answerType": "dropdown",
      "questionType": "standard",
      "options": [
        {
          "option": "Masculin ",
          "label": {
            "en": "Male",
            "fr": "Masculin"
          }
        },
        {
          "option": "Féminin",
          "label": {
            "en": "Female",
            "fr": "Féminin"
          }
        }
      ],
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Gender"
      }
    },
    {
      "name": "birthdate",
      "label": {
        "en": "5. What is your birth date?"
      },
      "answerType": "date",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Birth date"
      }
    },
    {
      "name": "documentid",
      "label": {
        "en": "6. What is your document ID number? (- of none)"
      },
      "answerType": "text",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "duplicateCheck": true,
      "phases": ["registrationValidation", "inclusion", "payment"],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Document ID"
      }
    },
    {
      "name": "headhousehold",
      "label": {
        "en": "7. Are you the head of household?"
      },
      "answerType": "dropdown",
      "questionType": "standard",
      "options": [
        {
          "option": "Oui",
          "label": {
            "en": "Yes",
            "fr": "Oui"
          }
        },
        {
          "option": "Non",
          "label": {
            "en": "No",
            "fr": "Non"
          }
        }
      ],
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Household head"
      }
    },
    {
      "name": "males",
      "label": {
        "en": "8. How many male(s) of 18 or older live in your household? (0 if none)"
      },
      "answerType": "numeric",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Men"
      }
    },
    {
      "name": "boys",
      "label": {
        "en": "9. How many boy(s) of 17 or younger live in your household? (0 if none)"
      },
      "answerType": "numeric",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "multiplier": 1
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Boys"
      }
    },
    {
      "name": "women",
      "label": {
        "en": "10. How many women of 18 or older live in your household? (0 if none)"
      },
      "answerType": "numeric",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Women"
      }
    },
    {
      "name": "girls",
      "label": {
        "en": "11. How many girl(s) of 17 or younger live in your household? (0 if none)"
      },
      "answerType": "numeric",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "multiplier": 1
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Girls"
      }
    },
    {
      "name": "householdsituation",
      "label": {
        "en": "12. What is your household's situation?"
      },
      "answerType": "dropdown",
      "questionType": "standard",
      "options": [
        {
          "option": "returnee",
          "label": {
            "en": "Retournee",
            "fr": "Retourné"
          }
        },
        {
          "option": "displaced",
          "label": {
            "en": "Displaced",
            "fr": "Déplacé"
          }
        },
        {
          "option": "autochtoneprivate",
          "label": {
            "en": "Autochtone (private residency)",
            "fr": "Autochtone (simple résident)"
          }
        },
        {
          "option": "autochtonehost",
          "label": {
            "en": "Autochtone (host family)",
            "fr": "Autochtone (famille d’accueil)"
          }
        }
      ],
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "returnee": 2,
        "displaced": 2,
        "autochtoneprivate": 0,
        "autochtonehost": 1
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Household situation"
      }
    },
    {
      "name": "maritalsituation",
      "label": {
        "en": "13. What is the marital situation of the head of your household?"
      },
      "answerType": "dropdown",
      "questionType": "standard",
      "options": [
        {
          "option": "widowed",
          "label": {
            "en": "Widowed",
            "fr": "Veuf/Veuve"
          }
        },
        {
          "option": "married",
          "label": {
            "en": "Married",
            "fr": "Marié(e)"
          }
        },
        {
          "option": "abandoned",
          "label": {
            "en": "Abandoned",
            "fr": "Abandonné(e)"
          }
        },
        {
          "option": "seperated",
          "label": {
            "en": "Seperated",
            "fr": "Séparé(e)"
          }
        },
        {
          "option": "divorced",
          "label": {
            "en": "Divorced",
            "fr": "Divorcé (e)"
          }
        },
        {
          "option": "girl",
          "label": {
            "en": "Girl/adolescent < 18 pregnant or mother",
            "fr": "Enfant/adolescent ménage"
          }
        }
      ],
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "widowed": 1,
        "married": 0,
        "abandoned": 1,
        "seperated": 1,
        "divorced": 1,
        "girl": 3
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Marital situation"
      }
    },
    {
      "name": "specialcharacteristics",
      "label": {
        "en": "14. What are special characteristics of the head of your household (if any)?"
      },
      "answerType": "multi-select",
      "questionType": "standard",
      "options": [
        {
          "option": "olderalone",
          "label": {
            "en": "Older person living alone"
          }
        },
        {
          "option": "elderlycharge",
          "label": {
            "en": "Elderly person in charge of a household"
          }
        },
        {
          "option": "pregnantbrestfeeding",
          "label": {
            "en": "Pregnant/breastfeeding woman"
          }
        },
        {
          "option": "womancharge",
          "label": {
            "en": "Woman in charge of a household"
          }
        },
        {
          "option": "childcharge",
          "label": {
            "en": "Child/adolescent in charge of a household"
          }
        },
        {
          "option": "girlmother",
          "label": {
            "en": "Girl/adolescent < 18 pregnant or mother"
          }
        },
        {
          "option": "none",
          "label": {
            "en": "None"
          }
        }
      ],
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "olderalone": 0,
        "elderlycharge": 1,
        "pregnantbrestfeeding": 2,
        "womancharge": 1,
        "childcharge": 3,
        "girlmother": 4,
        "none": 0
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Special characteristics"
      }
    },
    {
      "name": "pregnant",
      "label": {
        "en": "15. How many pregnant women are in your household? (0 if none)"
      },
      "answerType": "numeric",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "multiplier": 1
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Pregnant women"
      }
    },
    {
      "name": "breastfeeding",
      "label": {
        "en": "16. How many breastfeeding women are in your household? (0 if none)"
      },
      "answerType": "numeric",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "multiplier": 1
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Breastfeeding women"
      }
    },
    {
      "name": "womendisabled",
      "label": {
        "en": "17. How many women in your household aged 18 and over have a disability? (0 if none)"
      },
      "answerType": "numeric",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "multiplier": 2
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Women disability"
      }
    },
    {
      "name": "mendisabled",
      "label": {
        "en": "18. How many men in your household aged 18 and over have a disability? (0 if none)"
      },
      "answerType": "numeric",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "multiplier": 2
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Men disability"
      }
    },
    {
      "name": "girlsdisabled",
      "label": {
        "en": "19. How many girls in your household aged under 18 have a disability? (0 if none)"
      },
      "answerType": "numeric",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "multiplier": 2
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Girls disability"
      }
    },
    {
      "name": "boysdisabled",
      "label": {
        "en": "20. How many boys in your household aged under 18 have a disability? (0 if none)"
      },
      "answerType": "numeric",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "multiplier": 2
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Boys disability"
      }
    },
    {
      "name": "childrenmalnourished",
      "label": {
        "en": "21. How many malnourished children are in your household? (0 if none)"
      },
      "answerType": "numeric",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {
        "multiplier": 2
      },
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Malnourished children"
      }
    },
    {
      "name": "malnourishedtreatment",
      "label": {
        "en": "22. If there are malnourished children in your household are you pursuing treatment at the health center?"
      },
      "answerType": "dropdown",
      "questionType": "standard",
      "options": [
        {
          "option": "Oui",
          "label": {
            "en": "Yes",
            "fr": "Oui"
          }
        },
        {
          "option": "Non",
          "label": {
            "en": "No",
            "fr": "Non"
          }
        }
      ],
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Malnourished treatment"
      }
    }
  ],
  "aboutProgram": {
    "en": "[about aid program]"
  },
  "fullnameNamingConvention": ["name", "middlename", "firstname"],
  "languages": ["en"]
}
