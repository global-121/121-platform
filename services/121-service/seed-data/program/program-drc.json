{
  "published": true,
  "validation": true,
  "phase": "design",
  "location": "Democratic Republic of the Congo",
  "ngo": "Help a Child",
  "titlePortal": {
    "en": "Help a Child Digital Aid DRC"
  },
  "titlePaApp": {
    "en": "Help a Child DRC Cash Support"
  },
  "description": {
    "en": ""
  },
  "startDate": "2020-09-01T12:00:00Z",
  "endDate": "2022-11-30T12:00:00Z",
  "currency": "USD",
  "distributionFrequency": "once",
  "distributionDuration": 1,
  "fixedTransferValue": 53,
  "financialServiceProviders": [
    {
      "fsp": "VodaCash"
    }
  ],
  "targetNrRegistrations": 870,
  "tryWhatsAppFirst": false,
  "phoneNumberPlaceholder": "+24 300 00 00 00",
  "programCustomAttributes": [],
  "programQuestions": [
    {
      "name": "lastName",
      "label": {
        "en": "B.2. Name",
        "fr": "B.2. Nom de famille"
      },
      "answerType": "text",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true
    },
    {
      "name": "middleName ",
      "label": {
        "en": "B.3. Middle name",
        "fr": "B.3. Deuxième prénom"
      },
      "answerType": "text",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true
    },
    {
      "name": "firstName",
      "label": {
        "en": "B.4. First name",
        "fr": "B.4. Prénom"
      },
      "answerType": "text",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true
    }
  ],
  "aboutProgram": {
    "en": "This <strong>Help a Child DRC</strong> aid program was created to support people who are in need. Help a Child DRC will distribute money via VodaCom to included people.",
    "fr": "Le programme d'aide Help a Child DRC a été créé pour soutenir les personnes dans le besoin. Help a Child DRC distribuera de l'argent via VodaCom aux personnes concernées."
  },
  "fullnameNamingConvention": ["lastName", "middleName", "firstName"],
  "languages": ["en", "fr"],
  "enableMaxPayments": false,
  "enableScope": false,
  "allowEmptyPhoneNumber": false
}
