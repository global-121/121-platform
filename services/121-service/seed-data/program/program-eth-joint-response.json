{
  "published": false,
  "validation": false,
  "phase": "registrationValidation",
  "location": "Demoland",
  "ngo": "Demo NGO",
  "titlePortal": {
    "en": "Demo Cash Aid Project"
  },
  "titlePaApp": {
    "en": "Demo Cash Aid Project"
  },
  "description": {
    "en": ""
  },
  "startDate": "2022-09-29T12:00:00Z",
  "endDate": "2030-11-30T12:00:00Z",
  "currency": "ETB",
  "distributionFrequency": "month",
  "distributionDuration": 6,
  "fixedTransferValue": 1.0,
  "financialServiceProviders": [
    {
      "fsp": "Commercial-bank-ethiopia"
    }
  ],
  "targetNrRegistrations": 870,
  "tryWhatsAppFirst": false,
  "meetingDocuments": {
    "en": ""
  },
  "bankAccountNumberPlaceholder": "1000263499216",
  "programCustomAttributes": [],
  "programQuestions": [
    {
      "name": "fullname",
      "label": {
        "en": "What is your first name?"
      },
      "answerType": "text",
      "questionType": "standard",
      "options": null,
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "First name"
      }
    },
    {
      "name": "gender",
      "label": {
        "en": "4. What is your gender?"
      },
      "answerType": "dropdown",
      "questionType": "standard",
      "options": [
        {
          "option": "Masculin ",
          "label": {
            "en": "Male",
            "fr": "Masculin"
          }
        },
        {
          "option": "Féminin",
          "label": {
            "en": "Female",
            "fr": "Féminin"
          }
        }
      ],
      "persistence": true,
      "export": ["all-people-affected", "included", "selected-for-validation"],
      "scoring": {},
      "phases": [],
      "editableInPortal": true,
      "shortLabel": {
        "en": "Gender"
      }
    }
  ],
  "aboutProgram": {
    "en": "[about aid program]"
  },
  "fullnameNamingConvention": ["fullname"],
  "languages": ["en"],
  "enableMaxPayments": true
}
