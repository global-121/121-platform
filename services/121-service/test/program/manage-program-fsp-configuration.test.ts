/* eslint-disable jest/no-conditional-expect */
import { HttpStatus } from '@nestjs/common';

import {
  FspConfigurationProperties,
  Fsps,
} from '@121-service/src/fsps/enums/fsp-name.enum';
import { CreateProgramFspConfigurationDto } from '@121-service/src/program-fsp-configurations/dtos/create-program-fsp-configuration.dto';
import { UpdateProgramFspConfigurationDto } from '@121-service/src/program-fsp-configurations/dtos/update-program-fsp-configuration.dto';
import { UpdateProgramFspConfigurationPropertyDto } from '@121-service/src/program-fsp-configurations/dtos/update-program-fsp-configuration-property.dto';
import { RegistrationStatusEnum } from '@121-service/src/registration/enum/registration-status.enum';
import { SeedScript } from '@121-service/src/scripts/enum/seed-script.enum';
import { programIdVisa } from '@121-service/src/seed-data/mock/visa-card.data';
import { paymentNrVisa } from '@121-service/src/seed-data/mock/visa-card.data';
import programOCW from '@121-service/src/seed-data/program/program-nlrc-ocw.json';
import { getTransactions } from '@121-service/test/helpers/program.helper';
import {
  deleteProgramFspConfiguration,
  deleteProgramFspConfigurationProperty,
  getFspProgramProperties,
  getProgramFspConfigurations,
  patchProgramFspConfiguration,
  patchProgramFspConfigurationProperty,
  postProgramFspConfiguration,
  postProgramFspConfigurationProperties,
} from '@121-service/test/helpers/program-fsp-configuration.helper';
import {
  awaitChangeRegistrationStatus,
  deleteRegistrations,
  seedPaidRegistrations,
  waitForStatusChangeToComplete,
} from '@121-service/test/helpers/registration.helper';
import {
  getAccessToken,
  resetDB,
} from '@121-service/test/helpers/utility.helper';
import { registrationOCW5 } from '@121-service/test/registrations/pagination/pagination-data';

// Only tests most of the happy paths, edge cases are mostly covered in the unit tests

const seededFspConfigVoucher = programOCW.programFspConfigurations.find(
  (fspConfig) => fspConfig.fsp === Fsps.intersolveVoucherWhatsapp,
)!;

const createProgramFspConfigurationDto: CreateProgramFspConfigurationDto = {
  name: 'Intersolve Voucher WhatsApp name',
  label: {
    en: 'Intersolve Voucher WhatsApp label',
    nl: 'Intersolve Voucher WhatsApp label Dutch translation',
    es: 'Intersolve Voucher WhatsApp label Spanish translation',
  },
  fspName: Fsps.intersolveVoucherWhatsapp,
  properties: [
    {
      name: FspConfigurationProperties.username,
      value: 'user123',
    },
    {
      name: FspConfigurationProperties.password,
      value: 'password123',
    },
  ],
};

describe('Manage Fsp configurations', () => {
  let accessToken: string;

  beforeEach(async () => {
    await resetDB(SeedScript.nlrcMultiple, __filename);
    accessToken = await getAccessToken();
  });

  it('should add program Fsp configuration to an existing program', async () => {
    // Act
    const result = await postProgramFspConfiguration({
      programId: programIdVisa,
      body: createProgramFspConfigurationDto,
      accessToken,
    });
    const getResult = await getProgramFspConfigurations({
      programId: programIdVisa,
      accessToken,
    });
    const getResultConfig = getResult.body.find(
      (config) => config.name === createProgramFspConfigurationDto.name,
    );
    // Assert
    expect(result.statusCode).toBe(HttpStatus.CREATED);
    expect(result.body).toEqual(
      expect.objectContaining({
        name: createProgramFspConfigurationDto.name,
        label: createProgramFspConfigurationDto.label,
        fspName: createProgramFspConfigurationDto.fspName,
      }),
    );
    const propertyNamesResult = result.body.properties
      .map((property) => property.name)
      .sort();
    const propertyNamesExpected = createProgramFspConfigurationDto
      .properties!.map((property) => property.name)
      .sort();
    expect(propertyNamesResult).toEqual(
      expect.arrayContaining(propertyNamesExpected),
    );
    // All properties should have updated field as timestamp
    result.body.properties.forEach((property) => {
      const date = new Date(property.updated);
      expect(!isNaN(date.getTime())).toBeTruthy();
    });
    // Ensure that the update data is reflected in the get response so actually updated in the db
    expect(getResultConfig).toEqual(result.body);
  });

  it('should patch existing program Fsp configuration', async () => {
    // Act
    const updateProgramFspConfigurationDto: UpdateProgramFspConfigurationDto = {
      label: {
        en: 'Intersolve Voucher WhatsApp label updated',
        nl: 'Intersolve Voucher WhatsApp label Dutch translation updated',
        es: 'Intersolve Voucher WhatsApp label Spanish translation updated',
      },
      properties: [
        {
          name: FspConfigurationProperties.username,
          value: 'user1234',
        },
      ],
    };
    const name = seededFspConfigVoucher.fsp;
    const result = await patchProgramFspConfiguration({
      programId: programIdVisa,
      name,
      body: updateProgramFspConfigurationDto,
      accessToken,
    });
    const getResult = await getProgramFspConfigurations({
      programId: programIdVisa,
      accessToken,
    });
    const getResultConfig = getResult.body.find(
      (config) => config.name === name,
    );

    // Assert
    expect(result.statusCode).toBe(HttpStatus.OK);
    expect(result.body).toEqual(
      expect.objectContaining({
        name,
        label: updateProgramFspConfigurationDto.label,
        fspName: seededFspConfigVoucher.fsp,
      }),
    );
    const propertyNamesResult = result.body.properties.map(
      (property) => property.name,
    );
    const propertyNamesExpected =
      updateProgramFspConfigurationDto.properties!.map(
        (property) => property.name,
      );
    expect(propertyNamesResult).toEqual(
      expect.arrayContaining(propertyNamesExpected),
    );
    // All properties should have updated field as timestamp
    result.body.properties.forEach((property) => {
      const date = new Date(property.updated);
      expect(!isNaN(date.getTime())).toBeTruthy();
    });
    // Ensure that the update data is reflected in the get response so actually updated in the db
    expect(getResultConfig).toEqual(result.body);
  });

  it('should delete existing program Fsp configuration', async () => {
    // Act
    const name = seededFspConfigVoucher.fsp;
    const result = await deleteProgramFspConfiguration({
      programId: programIdVisa,
      name,
      accessToken,
    });
    const getResult = await getProgramFspConfigurations({
      programId: programIdVisa,
      accessToken,
    });
    const getResultConfig = getResult.body.find(
      (config) => config.name === name,
    );
    // Assert
    expect(result.statusCode).toBe(HttpStatus.NO_CONTENT);
    expect(getResultConfig).toBeUndefined();
  });

  it('should not delete existing program Fsp configuration because of active registrations with that config', async () => {
    // Prepare
    await seedPaidRegistrations([registrationOCW5], programIdVisa);

    // Act
    const name = seededFspConfigVoucher.fsp;
    const result = await deleteProgramFspConfiguration({
      programId: programIdVisa,
      name,
      accessToken,
    });
    const getResult = await getProgramFspConfigurations({
      programId: programIdVisa,
      accessToken,
    });
    const getResultConfig = getResult.body.find(
      (config) => config.name === name,
    );

    // Assert
    expect(result.statusCode).toBe(HttpStatus.CONFLICT);
    expect(result.body).toMatchSnapshot();
    expect(getResultConfig).toBeDefined();
  });

  // Checking this exception in api test because it's hard to unit test the more complex transaction querybuilder part
  it('deleting program Fsp configuration with existing transactions should set programFspConfigurationId of transactions to null', async () => {
    // Prepare
    await seedPaidRegistrations([registrationOCW5], programIdVisa);

    await awaitChangeRegistrationStatus({
      programId: programIdVisa,
      referenceIds: [registrationOCW5.referenceId],
      status: RegistrationStatusEnum.declined,
      accessToken,
    });
    await deleteRegistrations({
      programId: programIdVisa,
      referenceIds: [registrationOCW5.referenceId],
      accessToken,
    });
    await waitForStatusChangeToComplete(
      programIdVisa,
      1,
      RegistrationStatusEnum.deleted,
      8_000,
      accessToken,
    );

    // Act
    const name = seededFspConfigVoucher.fsp;
    const result = await deleteProgramFspConfiguration({
      programId: programIdVisa,
      name,
      accessToken,
    });
    const getResult = await getProgramFspConfigurations({
      programId: programIdVisa,
      accessToken,
    });
    const getResultConfig = getResult.body.find(
      (config) => config.name === name,
    );

    const getTranactions = await getTransactions({
      programId: programIdVisa,
      paymentNr: paymentNrVisa,
      registrationReferenceId: registrationOCW5.referenceId,
      accessToken,
    });

    // Assert
    expect(result.statusCode).toBe(HttpStatus.NO_CONTENT);
    expect(getResultConfig).not.toBeDefined();
    expect(getTranactions.body[0].programFspConfigurationName).toBe(null);
  });

  it('should add program Fsp configuration properties to an existing program Fsp configuration', async () => {
    // Prepare
    const createProgramFspConfigurationDtoNoProperties = {
      ...createProgramFspConfigurationDto,
      properties: undefined,
    };
    await postProgramFspConfiguration({
      programId: programIdVisa,
      body: createProgramFspConfigurationDtoNoProperties,
      accessToken,
    });

    // Act
    const result = await postProgramFspConfigurationProperties({
      programId: programIdVisa,
      properties: createProgramFspConfigurationDto.properties!,
      accessToken,
      name: createProgramFspConfigurationDto.name,
    });

    const getResult = await getProgramFspConfigurations({
      programId: programIdVisa,
      accessToken,
    });
    const getResultConfig = getResult.body.find(
      (config) => config.name === createProgramFspConfigurationDto.name,
    );

    // Assert
    expect(result.statusCode).toBe(HttpStatus.CREATED);
    const propertyNamesResult = result.body.map((property) => property.name);
    const propertyNamesExpected =
      createProgramFspConfigurationDto.properties!.map(
        (property) => property.name,
      );
    expect(propertyNamesResult).toEqual(
      expect.arrayContaining(propertyNamesExpected),
    );
    // All properties should have updated field as timestamp
    result.body.forEach((property) => {
      const date = new Date(property.updated);
      expect(!isNaN(date.getTime())).toBeTruthy();
    });
    // Ensure that the update data is reflected in the get response so actually updated in the db
    expect(getResultConfig?.properties.sort()).toEqual(result.body.sort());
  });

  it('should patch a property of an existing program Fsp configuration', async () => {
    // Prepare
    const updatedPropertyDto: UpdateProgramFspConfigurationPropertyDto = {
      value: 'user1234',
    };

    // Act
    const getResultBefore = await getProgramFspConfigurations({
      programId: programIdVisa,
      accessToken,
    });
    const usernamePropertyBefore = getResultBefore.body
      .find((config) => config.name === seededFspConfigVoucher.fsp)!
      .properties.find(
        (property) => property.name === FspConfigurationProperties.username,
      );

    const patchResult = await patchProgramFspConfigurationProperty({
      programId: programIdVisa,
      configName: seededFspConfigVoucher.fsp,
      propertyName: FspConfigurationProperties.username,
      body: updatedPropertyDto,
      accessToken,
    });

    const getResultAfter = await getProgramFspConfigurations({
      programId: programIdVisa,
      accessToken,
    });
    const usernamePropertyAfter = getResultAfter.body
      .find((config) => config.name === seededFspConfigVoucher.fsp)!
      .properties.find(
        (property) => property.name === FspConfigurationProperties.username,
      );

    // Assert
    expect(patchResult.statusCode).toBe(HttpStatus.OK);

    // Ensure that the username property value has been updated checking if the updated timestamp is later, since the value is not returned in the response
    expect(usernamePropertyAfter?.updated).not.toEqual(
      usernamePropertyBefore?.updated,
    );
    expect(new Date(usernamePropertyAfter!.updated).getTime()).toBeGreaterThan(
      new Date(usernamePropertyBefore!.updated).getTime(),
    );
  });

  it('should delete a property of an existing program Fsp configuration', async () => {
    // Act
    const deleteResult = await deleteProgramFspConfigurationProperty({
      programId: programIdVisa,
      configName: seededFspConfigVoucher.fsp,
      propertyName: FspConfigurationProperties.username,
      accessToken,
    });

    const getResultAfter = await getProgramFspConfigurations({
      programId: programIdVisa,
      accessToken,
    });
    const config = getResultAfter.body.find(
      (config) => config.name === seededFspConfigVoucher.fsp,
    );

    const usernamePropertyAfter = config?.properties.find(
      (property) => property.name === FspConfigurationProperties.username,
    );

    // Assert
    expect(deleteResult.statusCode).toBe(HttpStatus.NO_CONTENT);
    expect(usernamePropertyAfter).toBeUndefined();
    expect(config?.properties.length).toBe(
      seededFspConfigVoucher.properties.length - 1,
    );
  });

  it('Should return all visible properties of a program Fsp configuration', async () => {
    // Arrange
    const enumValues = Object.values(FspConfigurationProperties);
    // Act
    const getVisibleProperties = await getFspProgramProperties({
      programId: programIdVisa,
      configName: 'Intersolve-visa', //This configuration has visible properties
      accessToken,
    });
    // Assert
    expect(getVisibleProperties.statusCode).toBe(HttpStatus.OK);
    const properties = getVisibleProperties.body;
    properties.forEach((property) => {
      expect(property.value).not.toBe('[********]'); // Visible properties should not be masked
    });
    properties.forEach((property) => {
      expect(property.name).not.toBe('username');
      expect(property.name).not.toBe('password');
    });
    properties.forEach((property) => {
      expect(enumValues).toContain(property.name);
    });
  });

  it('Returns masked values for hidden properties of a program Fsp configuration', async () => {
    // Act
    const getHiddenProperties = await getFspProgramProperties({
      programId: programIdVisa,
      configName: 'Intersolve-voucher-whatsapp', // This configuration has hidden properties
      accessToken,
    });
    // Assert
    expect(getHiddenProperties.statusCode).toBe(HttpStatus.OK);
    const properties = getHiddenProperties.body;
    const hiddenPropertyNames = properties.map((property) => property.name);
    // Checks that only hidden properties are returned
    expect(hiddenPropertyNames).toEqual(
      expect.arrayContaining(['username', 'password']),
    );
    properties.forEach((property) => {
      expect(property.value).toBe('[********]'); // Hidden properties should be masked
    });
  });
});
