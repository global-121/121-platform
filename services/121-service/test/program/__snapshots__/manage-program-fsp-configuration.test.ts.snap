// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manage financial service provider configurations should not delete existing program financial service provider configuration because of active registrations with that config 1`] = `
{
  "message": "Cannot delete program financial service provider configuration Intersolve-voucher-whatsapp because it is still in use by registrations with referenceIds: 54e62864557597e034",
  "statusCode": 409,
}
`;
